### ISSUE TITLE
[Go] Build failure for 32-bit platform

### ISSUE THREAD

**Issue by jmacd** [apache/arrow#35861]
### Describe the bug, including details regarding any error messages, version, and platform.

With a dependency from https://github.com/f5/otel-arrow-adapter, for an OpenTelemetry implementation at https://github.com/lightstep/otel-launcher-go, I have this build failure in a step that tests with `GOARCH=386`:

```
Error: ../../../go/pkg/mod/github.com/apache/arrow/go/v12@v12.0.0-20230404000714-f02d35119ae6/arrow/array/dictionary.go:1556:22: math.MaxUint32 (untyped int constant 4294967295) overflows int
```

I expect the Arrow library to work on a 32-bit platform. 

### Component(s)

Go

**Comment by jmacd** [comment]
https://github.com/lightstep/otel-launcher-go/actions/runs/5138047739/jobs/9246939471

**Comment by zeroshade** [comment]
Hey @jmacd, this got addressed in #34784 and will be included in the v12.0.1 patch release.

**Comment by jmacd** [comment]
Thanks!
